### PR TITLE
feat: Add support for HCP Packer

### PR DESCRIPTION
**Pull request by sestegra**
## Summary of Pull Request

Add support for HCP Packer in Packer builds and usage examples with Terraform.

## Type of Pull Request

- [ ] This is a bugfix.
- [X] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is something else.
      Please describe:

## Context of the Pull Request

- New HCP Packer variable `common_hcp_packer_registry_enabled`
- Terraform examples for content-library and template items

## Related to Existing Issues

Closes: #231 

Related:  https://github.com/hashicorp/packer-plugin-vsphere/issues/207

## Test and Documentation Coverage

- [x] Tests have been completed (for bugfixes / features).
- [x] Documentation has been added / updated (for bugfixes / features).

## Breaking Changes?

- [ ] Yes, there are breaking changes.
- [X] No, there are no breaking changes.

### CODE REVIEW

File: builds/windows/desktop/10/windows.pkr.hcl [hunk 1]
```diff
@@ -209,4 +209,22 @@ build {
       vsphere_folder           = var.vsphere_folder
     }
   }
+
+  dynamic "hcp_packer_registry" {
+    for_each = var.common_hcp_packer_registry_enabled ? [1] : []
+    content {
+      bucket_name = replace("${var.vm_guest_os_family}-${var.vm_guest_os_name}-${var.vm_guest_os_version}-${var.vm_guest_os_edition}", ".", "")
```
Comment (tenthirtyam) [inline builds/windows/desktop/10/windows.pkr.hcl:216]: Should this be a local. e.g. `bucket_name = replace(local.bucket_name`...

File: builds/windows/desktop/10/windows.pkr.hcl [hunk 2]
```diff
@@ -209,4 +209,22 @@ build {
...
+      description = "${var.vm_guest_os_family} ${var.vm_guest_os_name} ${var.vm_guest_os_version} ${var.vm_guest_os_edition}"
```
Comment (tenthirtyam) [inline builds/windows/desktop/10/windows.pkr.hcl:217]: Should this be a local? e.g. `description = local.bucket_description`

File: terraform/vsphere-virtual-machine/content-library-ovf-hcp-packer/main.tf [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }
+  disk {
+    label            = "disk0"
+    size             = var.vm_disk_size
+    thin_provisioned = true
+  }
+  clone {
+    template_uuid = data.vsphere_content_library_item.content_library_item.id
+    customize {
+      linux_options {
```
Comment (tenthirtyam) [inline terraform/vsphere-virtual-machine/content-library-ovf-hcp-packer/main.tf:69]: May need to set the example directory name to Linux specific.

File: terraform/vsphere-virtual-machine/content-library-template-hcp-packer/main.tf [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  }
+  disk {
+    label            = "disk0"
+    size             = var.vm_disk_size
+    thin_provisioned = true
+  }
+  clone {
+    template_uuid = data.vsphere_content_library_item.content_library_item.id
+    customize {
+      linux_options {
```
Comment (tenthirtyam) [inline terraform/vsphere-virtual-machine/content-library-template-hcp-packer/main.tf:69]: May need to set the example directory name to Linux specific.

File: terraform/vsphere-virtual-machine/template-hcp-packer/main.tf [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  }
+  disk {
+    label            = "disk0"
+    size             = var.vm_disk_size
+    thin_provisioned = true
+  }
+  clone {
+    template_uuid = data.vsphere_virtual_machine.template.id
+    customize {
+      linux_options {
```
Comment (tenthirtyam) [inline terraform/vsphere-virtual-machine/template-hcp-packer/main.tf:65]: May need to set the example directory name to Linux specific.

File: build.sh [hunk 6]
```diff
@@ -725,6 +725,69 @@ menu_option_23() {
   echo "Done."
 }
 
+
+menu_option_24() {
+  INPUT_PATH="$SCRIPT_PATH"/builds/childs/template
+  echo -e "\nCONFIRM: Build a Ubuntu Server 22.04 LTS child from VM Template Template for VMware vSphere?"
```
Comment (tenthirtyam) [inline build.sh:731]: "Template Template"

File: build.sh [hunk 7]
```diff
@@ -725,6 +725,69 @@ menu_option_23() {
...
+  echo -e "\nContinue? (y/n)"
+  read -r REPLY
+  if [[ ! $REPLY =~ ^[Yy]$ ]]
+  then
+    exit 1
+  fi
+
+  ### Build a Ubuntu Server 22.04 LTS child from VM Template Template for VMware vSphere. ###
+  echo "Building a Ubuntu Server 22.04 LTS child from VM Template Template for VMware vSphere..."
```
Comment (tenthirtyam) [inline build.sh:740]: "Template Template"

File: build.yaml [hunk 8]
```diff
@@ -110,3 +110,18 @@ menu:
...
+      var_files:
+        - "vsphere.pkrvars.hcl"
+        - "build.pkrvars.hcl"
+        - "ansible.pkrvars.hcl"
+        - "proxy.pkrvars.hcl"
+        - "common.pkrvars.hcl"
+    submenu:
+      - entry: "Ubuntu Server 22.04 LTS child from VM Template"
+        build:
+          path: "builds/childs/template"
```
Comment (tenthirtyam) [inline build.yaml:124]: Should we consider `base` and `clone` terminology? Where base is the parent and clone is the child? If so, adjust the directory structure?

File: builds/windows/desktop/10/windows.pkr.hcl (hunk 1 above)
Comment (sestegra) [reply builds/windows/desktop/10/windows.pkr.hcl]: Done in commit f5d8ba04ffb9c2781c42aeb2ad231e419cf614b2

File: builds/windows/desktop/10/windows.pkr.hcl (hunk 2 above)
Comment (sestegra) [reply builds/windows/desktop/10/windows.pkr.hcl]: Done in commit f5d8ba04ffb9c2781c42aeb2ad231e419cf614b2

File: build.sh (hunk 6 above)
Comment (sestegra) [reply build.sh]: Done in commit 675af11a3a1d537610faac594556c4dea0ab05d4

File: build.sh (hunk 7 above)
Comment (sestegra) [reply build.sh]: Done in commit 675af11a3a1d537610faac594556c4dea0ab05d4

File: terraform/vsphere-virtual-machine/content-library-ovf-hcp-packer/main.tf (hunk 3 above)
Comment (sestegra) [reply terraform/vsphere-virtual-machine/content-library-ovf-hcp-packer/main.tf]: Done in commit 769325209ff3b4b6ae1d93e2262e7b32b449bf4d

File: terraform/vsphere-virtual-machine/content-library-template-hcp-packer/main.tf (hunk 4 above)
Comment (sestegra) [reply terraform/vsphere-virtual-machine/content-library-template-hcp-packer/main.tf]: Done in commit 769325209ff3b4b6ae1d93e2262e7b32b449bf4d

File: build.yaml (hunk 8 above)
Comment (sestegra) [reply build.yaml:124]: I prefer `base` and `clone` terminology.

Comment (sestegra) [reply build.yaml]: Done in commit d0f0f0646724ddca92b54c9566a9c33068f549dc

**Comment by tenthirtyam** [comment]
Thanks for the updates. I'll take another review and test next week.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Review by tenthirtyam — COMMENTED** [review]
I provided some minor refactoring to reduce the scope to basic configuration with HCP Packer, per our discussion over Slack.


🔧  Removes `/builds/clones*` (for now)
🔧  Updates `build.sh` and `build.yaml` to match the current upstream `main`.
🔧  Renames the path for the Terraform examples for HCP Packer consumption. May make some _minor_ adjustments for the `22.08` release.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Review by tenthirtyam — APPROVED** [review]
Approved! 🚀

**Review by tenthirtyam — APPROVED** [review]
LGTM! 🚀

**Pull request merged**